src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,311 @@
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { hol' := f.hol'.add g.hol',
+    bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).add (g.bdd_at_infty' A),
+    .. (f : slash_invariant_form Γ k) + g } }
+
+@[simp] lemma coe_add (f g : modular_form Γ k) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma add_apply (f g : modular_form Γ k) (z : ℍ) : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (modular_form Γ k) :=
+⟨{ to_fun := 0,
+  slash_action_eq' := slash_action.mul_zero _,
+  hol' := (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  bdd_at_infty' := λ A, by simpa [slash_action.mul_zero _ ]
+    using (bounded_at_im_infty_subalgebra ℂ).zero_mem' }⟩
+
+section
+variables {α : Type*} [has_smul α ℂ] [is_scalar_tower α ℂ ℂ]
+
+instance has_smul : has_smul α (modular_form Γ k) :=
+⟨ λ c f,
+  { to_fun := c • f,
+    hol' := by simpa using f.hol'.const_smul (c • (1 : ℂ)),
+    bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).const_smul_left (c • (1 : ℂ)),
+     .. c • (f : slash_invariant_form Γ k)}⟩
+
+@[simp] lemma coe_smul (f : (modular_form Γ k)) (n : α) : ⇑(n • f) = n • f := rfl
+@[simp] lemma smul_apply (f : (modular_form Γ k)) (n : α) (z : ℍ) :
+   (n • f) z = n • (f z) := rfl
+end
+
+instance has_neg : has_neg (modular_form Γ k) :=
+⟨λ f,
+  { to_fun := -f,
+    hol' := f.hol'.neg,
+    bdd_at_infty':= λ A, by simpa [-asymptotics.is_O_neg_left] using ((f.bdd_at_infty' A).neg_left),
+    .. -(f : slash_invariant_form Γ k) }⟩
+
+@[simp] lemma coe_neg (f : modular_form Γ k) : ⇑(-f) = -f := rfl
+
+@[simp] lemma neg_apply (f : modular_form Γ k) (z : ℍ) : (-f) z = - (f z) := rfl
+
+instance has_sub : has_sub (modular_form Γ k) :=
+⟨ λ f g, f + -g ⟩
+
+@[simp] lemma coe_sub (f g : (modular_form Γ k)) : ⇑(f - g) = f - g := rfl
+
+@[simp] lemma sub_apply (f g : modular_form Γ k) (z : ℍ) : (f - g) z = f z - g z := rfl
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl coe_add coe_neg coe_sub coe_smul coe_smul
+
+lemma coe_zero : ⇑(0 : modular_form Γ k) = (0 : ℍ → ℂ) := rfl
+
+/--Additive coercion from `modular_form` to `ℍ → ℂ`. -/
+def coe_hom : (modular_form Γ k) →+ (ℍ → ℂ) :=
+{ to_fun := λ f, f,
+  map_zero' := coe_zero,
+  map_add' := λ _ _, rfl }
+
+lemma coe_hom_injective : function.injective (@coe_hom Γ k) :=
+fun_like.coe_injective
+
+instance : module ℂ (modular_form Γ k) :=
+coe_hom_injective.module ℂ (coe_hom) (λ _ _, rfl)
+
+instance : inhabited (modular_form Γ k) := ⟨0⟩
+
+/--The modular form of weight `k_1 + k_2` given by the product of two modular forms of weights
+`k_1` and `k_2`. -/
+def mul {k_1 k_2 : ℤ} {Γ : subgroup SL(2, ℤ)} (f : (modular_form Γ k_1))
```

Comment:
I agree a graded ring instance would be good here, but I don't think it needs to be in this PR. Am I missing a comment above about it clashing with the fun_like stuff?

Reply:
> Am I missing a comment above about it clashing with the fun_like stuff?

@eric-wieser See https://github.com/leanprover-community/mathlib/pull/13250#issuecomment-1337107404.  It doesn't need to be (indeed shouldn't be) in this PR but the question is whether the setup is even compatible.  Does the `gring` setup allow for a disparate family of types, not all subtypes of a fixed type, to be a graded ring?

Reply:
Yes exactly, its the family of types which is what is worrying me now. I don't know how to prove a graded ring instance with these definitions.

Reply:
Yes,`gring` was explicitly designed to support families of types, but it's not all that easy to use.

Reply:
Which `gring` do you mean? `set_like.gring` or `direct_sum.gring`?

Reply:
The latter. The former is generally speaking easier to use, but if you're reasonable sure it makes sense to put `k` in the type, then `direct_sum.gring` is definitely viable.

The trick to proving it is to come up with some "obvious" map from `modular_form k` to `modular_form k'` when `k = k'`; which in your case is just "copy across the to_fun field and rewrite the proofs". You can then use that to eliminate the heqs that otherwise might appear

Reply:
Hmm I tryed using this map, but I'm stuggling to use it to remove the `heq`'s. Is there somewhere I can look at where this is done? I just dont know what `heq` lemma I should be using.

Reply:
I'm happy to have a go at doing this at some point soon. Do you have a branch I can play around withthat has the `heq`s you're stuck at?

The other example of how to use `gring` that's not in mathlib yet is #10255; but I think that one is more painful than what you need here.

Reply:
Ah ok #10255 was quite useful to look at. I just needed some `cast` lemmas to get it working. I think I can do this now. I wont add it to this PR but I'll make a separate one.